tools/file_packager.py

```diff
@@ -952,54 +952,62 @@ def generate_js(data_target, data_files, metadata):
-        xhr.onprogress = (event) => {
-          var url = packageName;
-          var size = packageSize;
-          if (event.total) size = event.total;
-          if (event.loaded) {
-            if (!xhr.addedTotal) {
-              xhr.addedTotal = true;
-              if (!Module['dataFileDownloads']) Module['dataFileDownloads'] = {};
-              Module['dataFileDownloads'][url] = {
-                loaded: event.loaded,
-                total: size
-              };
-            } else {
-              Module['dataFileDownloads'][url].loaded = event.loaded;
+        Module['dataFileDownloads'] ??= {};
+        fetch(packageName)
+          .catch((cause) => Promise.reject(new Error(`Network Error: ${packageName}`, {cause}))) // If fetch fails, rewrite the error to include the failing URL & the cause.
+          .then((response) => {
+
+            let loaded = 0;
+
+            if (!response.ok) {
+              return Promise.reject(new Error(`${response.status}: ${response.url}`));
             }
-            var total = 0;
-            var loaded = 0;
-            var num = 0;
-            for (var download in Module['dataFileDownloads']) {
-            var data = Module['dataFileDownloads'][download];
-              total += data.total;
-              loaded += data.loaded;
-              num++;
+
+            // If we're using the polyfill, readers won't be available...
+            if (!response.body && response.arrayBuffer) {
+              response.arrayBuffer().then(callback);
+              return;
             }
-            total = Math.ceil(total * Module['expectedDataFileDownloads']/num);
-            Module['setStatus']?.(`Downloading data... (${loaded}/${total})`);
-          } else if (!Module['dataFileDownloads']) {
+
+            const reader = response.body.getReader();
+            const headers = response.headers;
+
+            const size = Number(headers.get('Content-Length') ?? packageSize);
+            const chunks = [];
+
+            const iterate = () => reader.read().then(handleChunk).catch((cause) => {
+              return Promise.reject(new Error(`Unexpected error while handling : ${response.url} ${cause}`, {cause}));
+            });
+
+            const handleChunk = ({done, value}) => {
+              if (!done) {
+                chunks.push(value);
+                loaded += value.length;
+                Module['dataFileDownloads'][packageName] = {loaded, total: size};
+
+                let totalLoaded = 0;
+                let totalSize = 0;
+
+                for (const download of Object.values(Module['dataFileDownloads'])) {
+                  totalLoaded += download.loaded;
+                  totalSize += download.total;
+                }
+
+                Module['setStatus']?.(`Downloading data... (${totalLoaded}/${totalSize})`);
+                return iterate();
+              } else {
+                const packageData = new Uint8Array(chunks.reduce((a, b) => a.concat(b), []));
```

Comment:
@sbc100 Thanks for having me revisit this section. I eliminated the second pass.

Reply:
NICE!!!